liblwgeom/lwgeom_median.c

```diff
@@ -19,156 +19,247 @@
+	assert(input_ok != NULL);
+	assert(*input_ok == LW_TRUE);
+	assert(g->ngeoms > 0);
 	uint32_t i;
 	uint32_t n = 0;
-	int is_3d = lwgeom_has_z((LWGEOM*) g);
+	int has_z = lwgeom_has_z((LWGEOM*) g);
+	int has_m = lwgeom_has_m((LWGEOM*) g);
 
-	POINT3D* points = lwalloc(g->ngeoms * sizeof(POINT3D));
 	for (i = 0; i < g->ngeoms; i++)
 	{
+		assert(n < g->ngeoms);
+		assert(n <= i);
 		LWGEOM* subg = lwcollection_getsubgeom((LWCOLLECTION*) g, i);
 		if (!lwgeom_is_empty(subg))
 		{
-			getPoint3dz_p(((LWPOINT*) subg)->point, 0, (POINT3DZ*) &points[n++]);
-			if (!is_3d)
-				points[n-1].z = 0.0; /* in case the getPoint functions return NaN in the future for 2d */
+			*input_empty = LW_FALSE;
+			if (!getPoint4d_p(((LWPOINT*) subg)->point, 0, &points[n]))
+			{
+				lwerror("Geometric median: getPoint4d_p reported failure on point (POINT(%g %g %g %g), number %d of %d in input).", points[n].x, points[n].y, points[n].z, points[n].m, i, g->ngeoms);
+				*input_ok = LW_FALSE;
+				break;
+			}
+			if (has_m)
+			{
+				/* This limitation on non-negativity can be lifted
+				 * by replacing Weiszfeld algorithm with different one.
+				 * Possible option described in:
+				 *
+				 * Drezner, Zvi & O. Wesolowsky, George. (1991).
+				 * The Weber Problem On The Plane With Some Negative Weights.
+				 * INFOR. Information Systems and Operational Research.
+				 * 29. 10.1080/03155986.1991.11732158.
+				 */
+				if (points[n].m < 0)
+				{
+					lwerror("Geometric median input contains points with negative weights (POINT(%g %g %g %g), number %d of %d in input). Implementation can't guarantee global minimum convergence.", points[n].x, points[n].y, points[n].z, points[n].m, i, g->ngeoms);
+					*input_ok = LW_FALSE;
+					break;
+				}
+
+				/* points with zero weight are not going to affect calculation, drop them early */
+				if (points[n].m > DBL_EPSILON)
+				{
+					n++;
+				}
+			}
+			else
+			{
+				n++;
+			}
+		}
+	}
+	if (*input_ok)
+	{
+		if (!has_z)
+		{
+			for (i = 0; i < n; i++)
+			{
+				points[i].z = 0.0;
+			}
+		}
+		if (!has_m)
+		{
+			for (i = 0; i < n; i++)
+			{
+				points[i].m = 1.0;
+			}
 		}
 	}
 
-	if (ngeoms != NULL)
-		*ngeoms = n;
-
+	*npoints = n;
 	return points;
 }
 
+
 LWPOINT*
 lwmpoint_median(const LWMPOINT* g, double tol, uint32_t max_iter, char fail_if_not_converged)
 {
-	uint32_t npoints; /* we need to count this ourselves so we can exclude empties */
+	if (g->ngeoms == 0)
+	{
+		return lwpoint_construct_empty(g->srid, 0, 0);
+	}
+	assert(g->ngeoms > 0);
+
+	/* m ordinate is considered weight, if defined */
+	uint32_t npoints = 0; /* we need to count this ourselves so we can exclude empties and weightless points */
 	uint32_t i;
+	int input_ok = LW_TRUE;
+	int input_empty = LW_TRUE;
 	double delta = DBL_MAX;
-	double* distances;
-	POINT3D* points = lwmpoint_extract_points_3d(g, &npoints);
-	POINT3D median;
+	POINT4D median;
+	double* distances = NULL;
+	POINT4D* points = lwalloc(g->ngeoms * sizeof(POINT4D));
```

Comment:
Why is this being allocated here, and passed as an argument to `lwmpoint_extract_points_4d` (which then returns its argument???)

Reply:
rolled to previous syntax.